include/grpcpp/impl/codegen/message_allocator.h

```diff
@@ -22,31 +22,42 @@
 namespace grpc {
 namespace experimental {
 
-// This is per rpc struct for the allocator. We can potentially put the grpc
-// call arena in here in the future.
+// NOTE: This is an API for advanced users who need custom allocators.
+// Per rpc struct for the allocator. This is the interface to return to user.
+class RpcAllocatorState {
+ public:
+  virtual ~RpcAllocatorState() = default;
+  // Optionally deallocate request early to reduce the size of working set.
+  // A custom MessageAllocator needs to be registered to make use of this.
+  // This is not abstract because implementing it is optional.
+  virtual void FreeRequest() {}
+};
+
+// This is the interface returned by the allocator.
+// grpc library will call the methods to get request/response pointers and to
+// release the object when it is done.
 template <typename RequestT, typename ResponseT>
-struct RpcAllocatorInfo {
-  RequestT* request;
-  ResponseT* response;
-  // per rpc allocator internal state. MessageAllocator can set it when
-  // AllocateMessages is called and use it later.
-  void* allocator_state;
+class MessageHolder : public RpcAllocatorState {
+ public:
+  virtual void Release() { delete this; }
```

Comment:
Just out of curiosity, since this is returned by the allocator, any reason why it is not an abstract function?

Reply:
It seems a good default version, and thus saving most of the implementations a line or three... :)

Reply:
Should have a strong comment then since self-deletion is a somewhat big deal

Reply:
Just made it abstract and added a comment.